src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs

```diff
@@ -48,9 +54,19 @@ internal T[] Rent()
                 // Check if all of our buffers have been rented
                 if (_index < _data.Length)
                 {
-                    buffer = _data[_index] ?? new T[_bufferLength];
+                    buffer = _data[_index];
+                    if (buffer == null)
+                    {
+                        buffer = new T[_bufferLength];
+                        _eventSource.BufferCreated(buffer.GetHashCode(), _bufferLength, GetHashCode());
+                    }
                     _data[_index++] = null;
                 }
+                else if (_exhaustedEventSent == false)
+                {
+                    _eventSource.BucketExhausted(GetHashCode(), _bufferLength, _data.Length);
+                    _exhaustedEventSent = true;
+                }
```

Comment:
We only send it once per-exhausted state so we don't spam logs. If the bucket is continuously exhausted, it seems excessive to keep spamming the log; however, if it has a bunch of buffers returned and then exhausted again, we would send another log. 

I can change this if you'd like, I'm not tied to either implementation so if you think one way is better / fits more with the guidelines, I'll happily change it